controllers/s3/s3bucket.go

```diff
@@ -0,0 +1,244 @@
+	storagev1alpha1 "github.com/vshn/stack-cloudscale/api/storage/v1alpha1"
+	cloudscalev1alpha1 "github.com/vshn/stack-cloudscale/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/crossplaneio/crossplane-runtime/pkg/logging"
+	"github.com/crossplaneio/crossplane-runtime/pkg/meta"
+	"github.com/crossplaneio/crossplane-runtime/pkg/resource"
+	"github.com/vshn/stack-cloudscale/clients/s3"
+)
+
+const (
+	errNotInstance = "managed resource is not an S3Bucket"
+
+	statusOnline   = "Online"
+	statusCreating = "Creating"
+	statusDeleting = "Deleting"
+
+	resourceCredentialsSecretBucketname = "bucketname"
+)
+
+var log = logging.Logger.WithName("s3bucket_controller")
+
+// BucketInstanceController is responsible for adding the S3Bucket
+// controller and its corresponding reconciler to the manager with any runtime configuration.
+type BucketInstanceController struct{}
+
+// SetupWithManager instantiates a new controller using a resource.ManagedReconciler
+// configured to reconcile S3Buckets using an ExternalClient produced by
+// connecter, which satisfies the ExternalConnecter interface.
+func (r *BucketInstanceController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(strings.ToLower(storagev1alpha1.S3BucketKindAPIVersion)).
+		For(&storagev1alpha1.S3Bucket{}).
+		Owns(&corev1.Secret{}).
+		Complete(resource.NewManagedReconciler(mgr,
+			resource.ManagedKind(storagev1alpha1.S3BucketGroupVersionKind),
+			resource.WithExternalConnecter(&connecter{client: mgr.GetClient(), newS3Client: s3.NewClient})))
+}
+
+// Connecter satisfies the resource.ExternalConnecter interface.
+type connecter struct {
+	client      client.Client
+	newS3Client func(ctx context.Context, cloudscaleToken string, httpClient *http.Client) s3.Service
+}
+
+// Connect to the supplied resource.Managed (presumed to be a
+// S3Bucket) by using the Provider it references to create a new
+// S3 client.
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (resource.ExternalClient, error) {
+	// Assert that resource.Managed we were passed in fact contains an
+	// S3Bucket. We told NewControllerManagedBy that this was a
+	// controller For S3Bucket, so something would have to go
+	// horribly wrong for us to encounter another type.
+	i, ok := mg.(*storagev1alpha1.S3Bucket)
+	if !ok {
+		return nil, errors.New(errNotInstance)
+	}
+
+	// Get the Provider referenced by the S3Bucket.
+	p := &cloudscalev1alpha1.Provider{}
+	if err := c.client.Get(ctx, meta.NamespacedNameOf(i.Spec.ProviderReference), p); err != nil {
+		return nil, errors.Wrap(err, "cannot get Provider")
+	}
+
+	// Get the Secret referenced by the Provider.
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.Secret.Namespace, Name: p.Spec.Secret.Name}
+	if err := c.client.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrapf(err, "cannot get Provider secret %v", n.String())
```

Comment:
Nitpick: I'm pretty sure if you use `%s` instead of `%v` the `.String()` method will be implied.